apps/web/public/sw.js

```diff
@@ -0,0 +1,411 @@
+          req.onerror = () => reject(req.error);
+        });
+      try {
+        const result = await getData();
+        if (result.length)
+          return new Response(JSON.stringify({ labels: result }), {
+            status: 200,
+          });
+      } catch (error) {
+        console.log("Error while getting data from indexeddb", error.message);
+        // return new Response(`Error:  + ${error.message}`, {
+        //   status: 500,
+        //   headers: { "Content-Type": "text/plain" },
+        // });
+      }
+    }
+  }
+
+  if (
+    url.includes(".svg") ||
+    url.includes(".jpg") ||
+    url.includes(".png") ||
+    url.includes(".jpeg")
+  ) {
+    const cacheResponse = await caches.match(request);
+    if (cacheResponse) return cacheResponse;
+    const fetchResponse = await fetch(request);
+
+    //save in image cache
+    console.log(`save an IMAGE file ${url}`);
+    const cache = await caches.open(IMAGE_CACHE);
+    cache.put(request, fetchResponse.clone());
+    return fetchResponse;
+  }
+
+  if (url.includes("/api/user/stats/emails/all")) {
+    const requestData = await request.clone().json();
+    const localData = await loadLocalMail(!requestData.loadBefore);
+    if (localData?.length > 0) {
+      const newRequest = new Request(request, {
+        body: JSON.stringify({
+          ...requestData,
+          timestamp: localData[0].timestamp,
+        }),
+      });
+      // get Rest of the data
+      fetchAndUpdateMails(newRequest, !requestData.loadBefore);
+      return new Response(
+        JSON.stringify({
+          emails: localData,
+          page: Math.floor(localData.length / 100),
+        }),
+      );
+    }
+    const fetchResponse = await fetch(request);
+    const mails = await fetchResponse.json();
+
+    saveMails(mails);
+
+    return new Response(JSON.stringify(mails), { status: 200 });
+  }
+
+  if (url.includes("/api/user/stats/emails/sw")) {
+    const { searchParams } = new URL(url);
+    const params = {
+      period: searchParams.get("period") || "week", // "day", "week", "month", "year"
+      fromDate: searchParams.get("fromDate"), // null or undefined
+      toDate: searchParams.get("toDate"), // null or undefined
+    };
+    // 1. Get Data stats based on the timestamp
+    const stats = await getStatsByPeriod(params);
+    if (stats)
+      return new Response(JSON.stringify(stats), {
+        status: 200,
+      });
+
+    return new Response(
+      JSON.stringify({
+        message: "No stored data found in indexedDB for analytics",
+      }),
+    );
+  }
+  return fetch(request);
+}
+
+async function fetchAndUpdateIDB(request) {
+  const networkResponse = await fetch(request);
+  const clonedResponse = networkResponse.clone();
+
+  // open the database connection
+  const data = await clonedResponse.json();
+  if (!data?.labels) return;
+
+  await saveLabels(data.labels);
+  messageClients(MESSAGES.LABELS_UPDATED, data);
+}
+
+function messageClients(type, payload) {
```

Comment:
Broadcasts data to all the browser tabs linked to the service worker with the updated data.